<a href="https://colab.research.google.com/github/Guryansh/Fuel-Cell-Performance-Assignment/blob/main/ML_Assignmnet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install pycaret &> /dev/null

In [3]:
from pycaret.regression import *
import pandas as pd

In [5]:
data = pd.read_csv('Fuel_cell_performance_data-Full.csv')

In [17]:
selected_data = data[[data.columns[i] for i in range(15)]+['Target5']]
selected_data.head()
model_df = pd.DataFrame(columns=['Model Code', 'Model Parameters', 'R2'])

# **1.1 Model Using 10 Folds**

In [18]:
setup(data = selected_data, target='Target5', train_size = 0.7, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'None, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **1.2 Model Using 10 Folds - shuffle**

In [19]:
setup(data = selected_data, target='Target5', train_size = 0.7, data_split_shuffle = True, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'None - shuffle, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **1.3 Model Using 15 Folds**

In [20]:
setup(data = selected_data, target='Target5', train_size = 0.7, data_split_shuffle = False, verbose=False, fold=15)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'None, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **1.4 Model Using 15 Folds - shuffle**

In [21]:
setup(data = selected_data, target='Target5', train_size = 0.7, data_split_shuffle = True, verbose=False, fold=15)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'None - shuffle, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **2.1 Model performance using "Normalization" under 10 Folds**

In [22]:
setup(data = selected_data, target='Target5', train_size = 0.7,normalize = True, normalize_method = 'zscore', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Normalization, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **2.2 Model performance using "Normalization" under 15 Folds**

In [23]:
setup(data = selected_data, target='Target5', train_size = 0.7,normalize = True, normalize_method = 'zscore',fold=15, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Normalization, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **3.1 Model performance using "Feature Selection" under 10 Folds**

In [24]:
setup(data = selected_data, target='Target5', train_size = 0.7,feature_selection = True, feature_selection_method = 'classic',
      n_features_to_select = 0.2, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Feature Selection, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **3.2 Model performance using "Feature Selection" under 15 Folds**

In [25]:
setup(data = selected_data, target='Target5', train_size = 0.7, feature_selection = True, feature_selection_method = 'classic',
      n_features_to_select = 0.2, fold=15, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Feature Selection, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **4.1 Model performance using "Outlier Removal" under 10 Folds**

In [26]:
setup(data = selected_data, target='Target5', train_size = 0.7, remove_outliers = True, outliers_method = "iforest", outliers_threshold = 0.05, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Outlier Removal, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **4.2 Model performance using "Outlier Removal" under 15 Folds**

In [27]:
setup(data = selected_data, target='Target5', train_size = 0.7, remove_outliers = True, outliers_method = "iforest", outliers_threshold = 0.05, fold=15, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Outlier Removal, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **5.1 Model performance using "Transformation" under 10 Folds**

In [28]:
setup(data = selected_data, target='Target5', train_size = 0.7, transformation = True, transformation_method = 'yeo-johnson', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Transformation, 10 folds', r2_score]


Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **5.2 Model performance using "Transformation" under 15 Folds**

In [29]:
setup(data = selected_data, target='Target5', train_size = 0.7, transformation = True, transformation_method = 'yeo-johnson', fold=15, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'Transformation, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **6.1 Model performance using "PCA" under 10 Folds**

In [30]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'linear', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'PCA, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **6.2 Model performance using "PCA" under 15 Folds**

In [31]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'linear', fold=15, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'PCA, 15 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **Summary Before Next Combination**

In [32]:
display(model_df)

Model Code             Model Parameters      R2
0         ExtraTreesRegressor               None, 10 folds  0.7711
1         ExtraTreesRegressor     None - shuffle, 10 folds  0.7674
2               LGBMRegressor               None, 15 folds  0.7676
3   GradientBoostingRegressor               None, 15 folds  0.7543
4         ExtraTreesRegressor      Normalization, 10 folds  0.7677
5         ExtraTreesRegressor      Normalization, 15 folds  0.7732
6         ExtraTreesRegressor  Feature Selection, 10 folds  0.6909
7         ExtraTreesRegressor  Feature Selection, 15 folds  0.6936
8               LGBMRegressor    Outlier Removal, 10 folds  0.7671
9               LGBMRegressor    Outlier Removal, 15 folds  0.7621
10        ExtraTreesRegressor     Transformation, 10 folds  0.7704
11              LGBMRegressor     Transformation, 15 folds  0.7684
12              LGBMRegressor                PCA, 10 folds  0.7768
13              LGBMRegressor                PCA, 15 folds  0.7708

Based on Single Paramter selection we will choose combinations

# **7.1 Model Performance using Normalization and Outlier Removal**

In [33]:
setup(data = selected_data, target='Target5', train_size = 0.7, normalize = True, normalize_method = 'zscore', remove_outliers = True, outliers_method = "iforest", outliers_threshold = 0.05, data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'N+OR, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **7.2 Model Performance using Normalization and Transformation**

In [34]:
setup(data = selected_data, target='Target5', train_size = 0.7, normalize = True, normalize_method = 'zscore', transformation = True, transformation_method = 'yeo-johnson', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'N+T, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **7.3 Model Performance using Normalization and PCA**

In [35]:
setup(data = selected_data, target='Target5', train_size = 0.7, normalize = True, normalize_method = 'zscore', pca = True, pca_method = 'linear', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'N+PCA, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **7.4 Model Performance using Outlier Removal and PCA**

In [36]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'linear', remove_outliers = True, outliers_method = "iforest", outliers_threshold = 0.05,data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'OR+PCA, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **7.5 Model Performance using Outlier Removal and Transformation**

In [37]:
setup(data = selected_data, target='Target5', train_size = 0.7, remove_outliers = True, outliers_method = "iforest", outliers_threshold = 0.05, transformation = True, transformation_method = 'yeo-johnson', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'OR+T, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **7.6 Model Performance using PCA and Transformation**

In [38]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'linear', transformation = True, transformation_method = 'yeo-johnson', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'PCA+T, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **Summary of resulting combinations**

In [39]:
display(model_df)

Model Code             Model Parameters      R2
0         ExtraTreesRegressor               None, 10 folds  0.7711
1         ExtraTreesRegressor     None - shuffle, 10 folds  0.7674
2               LGBMRegressor               None, 15 folds  0.7676
3   GradientBoostingRegressor               None, 15 folds  0.7543
4         ExtraTreesRegressor      Normalization, 10 folds  0.7677
5         ExtraTreesRegressor      Normalization, 15 folds  0.7732
6         ExtraTreesRegressor  Feature Selection, 10 folds  0.6909
7         ExtraTreesRegressor  Feature Selection, 15 folds  0.6936
8               LGBMRegressor    Outlier Removal, 10 folds  0.7671
9               LGBMRegressor    Outlier Removal, 15 folds  0.7621
10        ExtraTreesRegressor     Transformation, 10 folds  0.7704
11              LGBMRegressor     Transformation, 15 folds  0.7684
12              LGBMRegressor                PCA, 10 folds  0.7768
13              LGBMRegressor                PCA, 15 folds  0.7708
14        ExtraTreesRegressor               N+OR, 10 folds  0.7606
15              LGBMRegressor                N+T, 10 folds  0.7680
16              BayesianRidge              N+PCA, 10 folds  0.7140
17        ExtraTreesRegressor             OR+PCA, 10 folds  0.7616
18              LGBMRegressor               OR+T, 10 folds  0.7619
19              LGBMRegressor              PCA+T, 10 folds  0.7696

So maximum coulbe achieved via PCA only so we shall focus on optimising it.

# **8.1 Model perfomance using PCA - kernel**

In [40]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'kernel', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'PCA - kernel, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **8.2 Model performance using PCA - incremental**

In [41]:
setup(data = selected_data, target='Target5', train_size = 0.7, pca = True, pca_method = 'incremental', data_split_shuffle = False, verbose=False)
cm = compare_models()
comparison_results = pull()
r2_score = comparison_results.loc[comparison_results.index[0], 'R2']
top_model = cm
model_df.loc[len(model_df)] = [top_model.__class__.__name__, 'PCA - incremental, 10 folds', r2_score]

Processing:   0%|          | 0/81 [00:00<?, ?it/s]

# **Result**

In [42]:
display(model_df)

Model Code             Model Parameters      R2
0         ExtraTreesRegressor               None, 10 folds  0.7711
1         ExtraTreesRegressor     None - shuffle, 10 folds  0.7674
2               LGBMRegressor               None, 15 folds  0.7676
3   GradientBoostingRegressor               None, 15 folds  0.7543
4         ExtraTreesRegressor      Normalization, 10 folds  0.7677
5         ExtraTreesRegressor      Normalization, 15 folds  0.7732
6         ExtraTreesRegressor  Feature Selection, 10 folds  0.6909
7         ExtraTreesRegressor  Feature Selection, 15 folds  0.6936
8               LGBMRegressor    Outlier Removal, 10 folds  0.7671
9               LGBMRegressor    Outlier Removal, 15 folds  0.7621
10        ExtraTreesRegressor     Transformation, 10 folds  0.7704
11              LGBMRegressor     Transformation, 15 folds  0.7684
12              LGBMRegressor                PCA, 10 folds  0.7768
13              LGBMRegressor                PCA, 15 folds  0.7708
14        ExtraTreesRegressor               N+OR, 10 folds  0.7606
15              LGBMRegressor                N+T, 10 folds  0.7680
16              BayesianRidge              N+PCA, 10 folds  0.7140
17        ExtraTreesRegressor             OR+PCA, 10 folds  0.7616
18              LGBMRegressor               OR+T, 10 folds  0.7619
19              LGBMRegressor              PCA+T, 10 folds  0.7696
20          AdaBoostRegressor       PCA - kernel, 10 folds -0.0099
21        ExtraTreesRegressor  PCA - incremental, 10 folds  0.7680

Best Case - Light Gradient Boosting Machine Regressor, PCA (linear) -> 0.7768


Worst Case - AdaBoost Regressor, PCA (kernel) -> -0.0099